app/main/forms/fields.py

```diff
@@ -0,0 +1,78 @@
+'''
+Fields for WTForms used in the Digital Marketplace frontend
+'''
+
+from wtforms import Field, Form, FormField, RadioField
+from wtforms.fields import IntegerField
+from wtforms.utils import unset_value
+
+import datetime
+
+
+class DMFormField(FormField):
+    '''
+    A wrapper for `wtforms.FormField`
+    '''
+    @property
+    def errors(self):
+        '''The validation errors, in a format suitable for frontend-toolkit'''
+        return [e for errors in self.form.errors.values() for e in errors]
+
+
+class DMRadioField(RadioField):
+    '''
+    A wrapper for `wtforms.RadioField`
+    '''
+    @property
+    def options(self):
+        '''The RadioField choices, in a format suitable for the frontend toolkit'''
+        return [{"label": label, "value": value} for value, label in self.choices]
+
+
+class DateField(Field):
+    '''
+    A field enclosure for a date entry fieldset
+
+    >>> from wtforms import Form
+    >>> from werkzeug.datastructures import MultiDict
+    >>> class DateForm(Form):
+    ...     date = DateField()
+    >>> form = DateForm(MultiDict({
+    ...     'date-day': '31',
+    ...     'date-month': '12',
+    ...     'date-year': '1999'}))
+    >>> form.date.data
+    datetime.date(1999, 12, 31)
+    '''
+
+    class _DateForm(Form):
+        day = IntegerField("Day")
+        month = IntegerField("Month")
+        year = IntegerField("Year")
+
+    def __init__(self, label=None, validators=None, separator='-', **kwargs):
+        super().__init__(label=label, validators=validators, **kwargs)
+        self.form_field = FormField(self._DateForm, separator=separator, **kwargs)
+
+    def _value(self):
+        if self.raw_data:
+            return self.raw_data[0]
+        else:
+            return {}
+
+    def process(self, formdata, data=unset_value):
+        self.process_errors = []
+
+        self.form_field.process(formdata, data)
+
+        raw_data = {field.name: field.raw_data[0] for field in self.form_field
+                    if field.raw_data}
+        if not any(raw_data.values()):
+            raw_data = {}
+        self.raw_data = [raw_data]
```

Comment:
I may have misunderstood, but I think you need to make sure that `raw_data` only gets set if `formdata is not None`.

Also you might be missing a check on whether `data is unset_value` - or maybe not, perhaps that case just falls out. But I was really just looking to see how similar this implementation is to the base class impl, and it seems slightly more different than I expected :)

I think your `if not any(raw_data.values())` test is unnecessary too - I can't see how a dict comprehension can ever not return a dict, even if it is empty?

Reply:
Would it be worth adding some comments to this logic? It's been through several iterations of TDD to get it to a state where it works in most of the common scenarios I could think of (with different combinations of validators), which is why it has a couple of weird quirks.

Reply:
Sure.

BTW I see now that your `any` test is converting a dict consisting only of empty strings into an empty dict, so I guess I can definitely see where I was wrong there.